feat: add state history doc (#372)
spacelift-io/user-documentation: docs/vendors/terraform/state-management.md

The state server is an HTTP endpoint implementing the Terraform [standard state management protocol](https://www.terraform.io/docs/backends/types/http.html){: rel="nofollow"}. Our backend always ensures that the credentials belong to one of the runs or tasks that are currently marked as active on our end, and their state indicates that they should be accessing or modifying the state. Once this is established, we just pass the request to S3 with the right parameters.

## State history

![State history tab](../../assets/screenshots/terraform/state-management/state-history-tab.png)

> If your state is managed by spacelift, you could list all the changes on your state, and eventually rollback to an old version if needed.

Reviewer: ```suggestion
If your state is managed by spacelift, you can list all the changes on your state, and eventually rollback to an old version if needed.
```

![State history list](../../assets/screenshots/terraform/state-management/state-history-list.png)

!!! info

>     Not all runs or task will trigger a new state version, so you should not expect to see an exhaustive list of your runs and tasks in this list.

Reviewer: ```suggestion
    Not all runs or tasks will trigger a new state version, so you should not expect to see an exhaustive list of your runs and tasks in this list.
```

Also, I'd be tempted to add another sentence like "For example runs that produce no Terraform changes do not result in a new state version being created".

I know it seems obvious, but I think if we include a statement like this it's good to explicitly explain why.

Author: Yep, it can't hurt to be over explicit 👍🏻

States versions are currently kept for **30 days**.

### State rollback

> In some unusual scenario you could end with a broken or corrupted state to be created. That could happen for example if there is a bug during a terraform provider upgrage.

Reviewer: ```suggestion
In certain unusual scenarios you can end up with a broken or corrupted state being created. This could happen for example if there was a bug during a Terraform provider upgrade.
```

> State rollback allow you to rollback your state to a previous version.

Reviewer: ```suggestion
State rollback allows you to recover from this by rolling back your state to a previous version.
```

> Rolling-back a state will **not change any bit of your current infrastructure**, it's just reverting a file to an old version.

Reviewer: ```suggestion
Rolling back your state will **not apply any changes to your current infrastructure**. It just reverts your state file to an older version.
```

> It's up to the user to trigger proper tasks or runs to fix the state and re-apply the desired terraform configuration.

Reviewer: ```suggestion
It's up to you to trigger the proper tasks or runs to fix the state and re-apply the desired Terraform configuration.
```

Just for consistency - I'd say we either use the word "you", and write the article as though we're having a conversation with someone (which I personally like), or we need to use "the user" everywhere.

Author: Yeah makes perfect sense 👍🏻

!!! warning
    You should really understand what you are doing when performing a rollback.

>     State rollback should be used as a breaking glass operation just after a corrupted state has been created.

Reviewer: ```suggestion
    State rollback should be used as a break-glass operation just after a corrupted state has been created.
```

> If a stack is currenty using a rolled-back state, a warning will be shown in stack header.

Reviewer: ```suggestion
If a stack is currently using a rolled-back state, a warning will be shown in the stack header.
```

![stack with current state rollback](../../assets/screenshots/terraform/state-management/state-rollback-stack-header.png)

> To be able to roll back a state, the 3 conditions below should be satisfied:

Reviewer: ```suggestion
To be able to roll back a state, the 3 conditions below must be satisfied:
```

- You must be a stack admin
- The stack must be locked

> - The state should not have any pending runs or tasks

Reviewer: ```suggestion
- The state must not have any pending runs or tasks
```

> When those 3 conditions are met, you should be able to rollback your stack to a previous version of your state file.

Reviewer: ```suggestion
If those three conditions are met, you will be able to rollback your stack to a previous version of your state file.
```

![state rollback action](../../assets/screenshots/terraform/state-management/state-rollback-action.png)

> A new version of your state will be created on top of the others state versions and will be marked as a rollback

Reviewer: ```suggestion
After rollback completes successfully, a new version of your state will appear above the other state versions and will be marked as a rollback.
```

The state server is an HTTP endpoint implementing the Terraform [standard state management protocol](https://www.terraform.io/docs/backends/types/http.html){: rel="nofollow"}. Our backend always ensures that the credentials belong to one of the runs or tasks that are currently marked as active on our end, and their state indicates that they should be accessing or modifying the state. Once this is established, we just pass the request to S3 with the right parameters.

## State history

![State history tab](../../assets/screenshots/terraform/state-management/state-history-tab.png)

If your state is managed by spacelift, you can list all the changes on your state, and eventually rollback to an old version if needed.

![State history list](../../assets/screenshots/terraform/state-management/state-history-list.png)

!!! info
    Not all runs or tasks will trigger a new state version, so you should not expect to see an exhaustive list of your runs and tasks in this list.
    For example runs that produce no Terraform changes do not result in a new state version being created.

> States versions are currently kept for **30 days**.

Reviewer: Just thinking, maybe we should say something like `Non-current state versions are kept for **30 days**.`

It seems obvious, but just to make it clearer that it's only versions that aren't currently in use that are removed :laughing:
